### PR TITLE
ADD auto backport functionality to AD

### LINKED ISSUE #330: Add auto backport functionality to backport PRs for releases
Coming from meta issue [opensearch-project/opensearch-plugins#108](https://github.com/opensearch-project/opensearch-plugins/issues/108):

- [ ] Add auto-backport PR creation Github action to make sure all the PRs are backported for a particular release. Related documentation: https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md.
- [ ] Add documentation for the same.

### PR DESCRIPTION
**Pull request by amitgalitz**
Signed-off-by: Amit Galitzky <amgalitz@amazon.com>

### Description
Adds backport workflow to allow automatic backport PR creation as we merge to main or label PR. Explanation of script added to documentation.

#### Side Notes:
 * We can label previously closed PRs and the action will run.
 * As of right now we have to manually delete the branch that the bot creates. This could be changed in the future to be automatic.

### Issues Resolved
resolves #330 
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -79,4 +79,12 @@ Sometimes you need to launch a cluster with more than one OpenSearch server proc
...
 You can also debug a multi-node cluster, by using a combination of above multi-node and debug steps.
-But, you must set up debugger configurations to listen on each port starting from `5005` and increasing by 1 for each node.
\ No newline at end of file
+But, you must set up debugger configurations to listen on each port starting from `5005` and increasing by 1 for each node.
+
+### Backports
+
+The Github workflow in [`backport.yml`](.github/workflows/backport.yml) creates backport PRs automatically when the
+original PR with an appropriate label `backport <backport-branch-name>` is merged to main with the backport workflow
+run successfully on the PR. For example, if a PR on main needs to be backported to `1.x` branch, add a label
```
Comment (ylwu-amzn) [inline DEVELOPER_GUIDE.md:88]: Is it possible to add multiple labels to backport to multiple branches, like `1.x`, `1.2` ?

Comment (amitgalitz) [reply DEVELOPER_GUIDE.md:88]: Yes multiple backports also work. Adding multiple tags will create multiple PRs for each branch.

**Review by ylwu-amzn — APPROVED** [review]

**Comment by ohltyler** [comment]
> As of right now we have to manually delete the branch that the bot creates. This could be changed in the future to be automatic.

What branches will the bot create?

**Comment by amitgalitz** [comment]
> > As of right now we have to manually delete the branch that the bot creates. This could be changed in the future to be automatic.
> 
> What branches will the bot create?

Bot will create a new branch in order to backport PR from. The PR will be merging from a branch named `backport-<PR ID number>-to-<backport branch>` that the workflow created automatically with contents of whatever needs to be backported. This is an example of a workflow created backport PR: https://github.com/VachaShah/TestGithubActions/pull/7

**Comment by ohltyler** [comment]
> > > As of right now we have to manually delete the branch that the bot creates. This could be changed in the future to be automatic.
> > 
> > 
> > What branches will the bot create?
> 
> Bot will create a new branch in order to backport PR from. The PR will be merging from a branch named `backport-<PR ID number>-to-<backport branch>` that the workflow created automatically with contents of whatever needs to be backported. This is an example of a workflow created backport PR: [VachaShah/TestGithubActions#7](https://github.com/VachaShah/TestGithubActions/pull/7)

I see - should we use the default action, or refer to @VachaShah 's custom one which is being used in core OpenSearch ([link](https://github.com/opensearch-project/OpenSearch/pull/1862/files)), that allows for custom branch naming?

Also, @VachaShah , have you researched any auto-branch-deletion strategy? For example, something like https://github.com/dawidd6/action-delete-branch that could be ran when the backport PRs are merged?

**Comment by VachaShah** [comment]
> > > > As of right now we have to manually delete the branch that the bot creates. This could be changed in the future to be automatic.
> > > 
> > > 
> > > What branches will the bot create?
> > 
> > 
> > Bot will create a new branch in order to backport PR from. The PR will be merging from a branch named `backport-<PR ID number>-to-<backport branch>` that the workflow created automatically with contents of whatever needs to be backported. This is an example of a workflow created backport PR: [VachaShah/TestGithubActions#7](https://github.com/VachaShah/TestGithubActions/pull/7)
> 
> I see - should we use the default action, or refer to @VachaShah 's custom one which is being used in core OpenSearch ([link](https://github.com/opensearch-project/OpenSearch/pull/1862/files)), that allows for custom branch naming?
> 
> Also, @VachaShah , have you researched any auto-branch-deletion strategy? For example, something like https://github.com/dawidd6/action-delete-branch that could be ran when the backport PRs are merged?

I am working on adding VachaShah/backport to the OpenSearch repo. I am currently testing out the permissions so haven't updated the documentation yet till I get it working for the OpenSearch repo. Using the one with custom branch naming would allow to maintain proper branching permissions. 

I will add the branch deletion feature as well in VachaShah/backport once this is done.

**Comment by ohltyler** [comment]
> > > As of right now we have to manually delete the branch that the bot creates. This could be changed in the future to be automatic.
> > 
> > 
> > What branches will the bot create?
> 
> Bot will create a new branch in order to backport PR from. The PR will be merging from a branch named `backport-<PR ID number>-to-<backport branch>` that the workflow created automatically with contents of whatever needs to be backported. This is an example of a workflow created backport PR: [VachaShah/TestGithubActions#7](https://github.com/VachaShah/TestGithubActions/pull/7)

I see, seems this can get messy 

> 

Great, thanks! If you are working on adding auto-deletion in that package, then we can wait before implementing something ourselves. @amitgalitz I'm ok with these changes to merge for now, and we can update to add custom naming + auto deletion later

**Review by ohltyler — APPROVED** [review]

**Comment by amitgalitz** [comment]
> , then we can wait before implementing something ourselves. @amitgalitz I'm ok with these cha

Sounds good, I'll merge this now and I'll keep up to date for any updates.

**Comment by ohltyler** [comment]
Failing to backport since this was merged via merge commit. That option is now blocked in the repo. Manually merged the commit in `1.x` via #388

**Pull request merged**